sumpy/codegen.py

```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
             else:
                 first_group.append(child)
 
+        new_children = tuple(first_group + second_group)
+        if len(new_children) == len(expr.children) and \
+                all(child is orig_child for child, orig_child in
+                    zip(new_children, expr.children)):
+            return expr
         return prim.Sum(tuple(first_group+second_group))
 
     map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
 
 # }}}
 
 
-class MathConstantRewriter(CSECachingMapperMixin, IdentityMapper):
-    def map_variable(self, expr):
+class MathConstantRewriter(CSECachingMapperMixin, CallExternalRecMapper):
+    def map_variable(self, expr, *args):
         if expr.name == "pi":
             return prim.Variable("M_PI")
         else:
-            return IdentityMapper.map_variable(self, expr)
+            return expr
 
     map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
 
 
+def combine_mappers(*mappers):
```

Comment:
It'd be nice if all the mapper-combining stuff were gathered together in one `{{{` section.